src/qutip_qip/circuit/base_renderer.py

```diff
@@ -0,0 +1,212 @@
+        The default is None.
+
+    wire_label : Optional[List], optional
+        The labels of the wires. The default is None.
+
+    wire_color : Optional[str], optional
+        The color of the wires. The default is None.
+    """
+
+    dpi: int = 150
+    fontsize: int = 10
+    end_wire_ext: int = 2
+    padding: float = 0.3
+    gate_margin: float = 0.15
+    wire_sep: float = 0.5
+    layer_sep: float = 0.5
+    gate_pad: float = 0.05
+    label_pad: float = 0.1
+    fig_height: Optional[float] = None
+    fig_width: Optional[float] = 10
+    bulge: Union[str, bool] = True
+    align_layer: bool = False
+    theme: Optional[Union[str, Dict]] = "qutip"
+    title: Optional[str] = None
+    bgcolor: Optional[str] = None
+    color: Optional[str] = None
+    wire_label: Optional[List] = None
+    wire_color: Optional[str] = None
+
+    def __post_init__(self):
+        if isinstance(self.bulge, bool):
+            self.bulge = "round4" if self.bulge else "square"
+
+        self.measure_color = "#000000"
+        if self.theme == "qutip":
+            self.theme = qutip
+        elif self.theme == "light":
+            self.theme = light
+        elif self.theme == "dark":
+            self.theme = dark
+            self.measure_color = "#FFFFFF"
+        elif self.theme == "modern":
+            self.theme = modern
+        else:
+            raise ValueError(
+                f"""Invalid theme: {self.theme},
+                Must be selectec from 'qutip', 'light', 'dark', or 'modern'.
+                """
+            )
+
+        self.bgcolor = self.bgcolor or self.theme["bgcolor"]
+        self.color = self.color or self.theme["color"]
+        self.wire_color = self.wire_color or self.theme["wire_color"]
+
+
+class BaseRenderer:
+    """
+    Base class for rendering quantum circuits with MatRender and TextRender.
+    """
+
+    def __init__(self, style: StyleConfig):
+        """
+        Initialize the base renderer with default values.
+        Both Renderers should override these attributes as needed.
+        """
+
+        self._qwires = 0
+        self._layer_list = []
+        self.style = style
+
+    def _get_xskip(self, wire_list: List[int], layer: int) -> float:
+        """
+        Get the xskip (horizontal value for getting to requested layer) for the gate to be plotted.
+
+        Parameters
+        ----------
+        wire_list : List[int]
+            The list of wires the gate is acting on (control and target).
+
+        layer : int
+            The layer the gate is acting on.
+
+        Returns
+        -------
+        float
+            The maximum xskip value needed to reach the specified layer.
+        """
+
+        if self.style.align_layer:
+            wire_list = list(range(self._qwires))
+
+        xskip = []
+        for wire in wire_list:
+            xskip.append(sum(self._layer_list[wire][:layer]))
+
+        return max(xskip)
```

Comment:
Does `aligh_layer` also work for the text renderer?

Reply:
I guess text rendering is by default aligned.

Reply:
Yes, it does perfectly 😃

Reply:
> I guess text rendering is by default aligned.

It’s not enabled by default—we still need to set the `align_layer=True` argument. This is essential because we’re continuing to use the layer concept to store gates at different horizontal positions, aiming for maximum compactness. The only change is that the minimum shift in the case of text is now 1 unit, rather than a floating-point value.